File: src/utils/pdf_report_builder.py
#!/usr/bin/env python3
"""
Build PDF reports for completed works using reportlab.
"""

import io
from collections import defaultdict
from datetime import datetime
from pathlib import Path

from reportlab.lib import colors
from reportlab.lib.enums import TA_CENTER, TA_LEFT, TA_RIGHT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.platypus import HRFlowable, Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle

from utils.db_storage import default_reports_dir
from utils.report_builder import _format_money


_PROJECT_ROOT = Path(__file__).resolve().parent.parent.parent
_DEFAULT_FONT_DIR = _PROJECT_ROOT / "resources" / "fonts"
_DEFAULT_FONT_FILE = _DEFAULT_FONT_DIR / "DejaVuSans.ttf"
_REGISTERED_FONT_NAME = "WorkOrderReportFont"

_SYSTEM_FONT_CANDIDATES = (
    "/System/Library/Fonts/Supplemental/Arial Unicode.ttf",
    "/Library/Fonts/Arial Unicode.ttf",
    "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
    "/usr/share/fonts/TTF/DejaVuSans.ttf",
    "C:/Windows/Fonts/arial.ttf",
)


def resolve_report_font_path(font_path=""):
    """Return path to TTF font that supports Cyrillic."""
    if font_path:
        path = Path(font_path).expanduser().resolve()
        if path.is_file():
            return path

    if _DEFAULT_FONT_FILE.is_file():
        return _DEFAULT_FONT_FILE

    for candidate in _SYSTEM_FONT_CANDIDATES:
        path = Path(candidate)
        if path.is_file():
            return path

    raise FileNotFoundError(
        "Report font not found. Place DejaVuSans.ttf in resources/fonts/ "
        "or install a system font with Cyrillic support."
    )


def register_report_font(font_path=""):
    """Register TTF font for reportlab and return registered name."""
    path = resolve_report_font_path(font_path)
    if _REGISTERED_FONT_NAME not in pdfmetrics.getRegisteredFontNames():
        pdfmetrics.registerFont(TTFont(_REGISTERED_FONT_NAME, str(path)))
    return _REGISTERED_FONT_NAME


class WorkReportPdfBuilder:
    """Build PDF work reports with the same content as plain-text reports."""

    _TABLE_COL_WIDTHS = [80 * mm, 20 * mm, 15 * mm, 30 * mm, 29 * mm]
    _META_COL_WIDTHS = [87 * mm, 87 * mm]

    def __init__(self, font_path=""):
        self._font_name = register_report_font(font_path)
        self._styles = self._build_styles()

    def _build_styles(self):
        styles = getSampleStyleSheet()
        base_kwargs = {"fontName": self._font_name}

        return {
            "title": ParagraphStyle(
                "ReportTitle",
                parent=styles["Heading1"],
                alignment=TA_CENTER,
                fontSize=14,
                leading=18,
                spaceAfter=6,
                **base_kwargs,
            ),
            "section": ParagraphStyle(
                "ReportSection",
                parent=styles["Heading2"],
                fontSize=11,
                leading=15,
                spaceBefore=8,
                spaceAfter=4,
                **base_kwargs,
            ),
            "normal": ParagraphStyle(
                "ReportNormal",
                parent=styles["Normal"],
                fontSize=10,
                leading=14,
                **base_kwargs,
            ),
            "table_header": ParagraphStyle(
                "ReportTableHeader",
                parent=styles["Normal"],
                alignment=TA_CENTER,
                fontSize=10,
                leading=12,
                spaceBefore=0,
                spaceAfter=0,
                **base_kwargs,
            ),
            "table_cell": ParagraphStyle(
                "ReportTableCell",
                parent=styles["Normal"],
                alignment=TA_LEFT,
                fontSize=10,
                leading=12,
                spaceBefore=0,
                spaceAfter=0,
                **base_kwargs,
            ),
            "table_total": ParagraphStyle(
                "ReportTableTotal",
                parent=styles["Normal"],
                alignment=TA_RIGHT,
                fontSize=11,
                leading=14,
                spaceBefore=0,
                spaceAfter=0,
                **base_kwargs,
            ),
        }

    def _table_style(self, include_total_row):
        commands = [
            ("GRID", (0, 0), (-1, -1), 0.5, colors.grey),
            ("BACKGROUND", (0, 0), (-1, 0), colors.lightgrey),
            ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
            ("TOPPADDING", (0, 0), (-1, -1), 3),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 3),
            ("ALIGN", (0, 0), (-1, 0), "CENTER"),
            ("ALIGN", (0, 1), (0, -2 if include_total_row else -1), "LEFT"),
            ("ALIGN", (1, 1), (-1, -2 if include_total_row else -1), "RIGHT"),
        ]
        if include_total_row:
            commands.extend([
                ("BACKGROUND", (0, -1), (-1, -1), colors.whitesmoke),
                ("ALIGN", (0, -1), (3, -1), "RIGHT"),
                ("SPAN", (0, -1), (3, -1)),
            ])
        return TableStyle(commands)

    def _work_table_header(self):
        return [
            Paragraph("Работа", self._styles["table_header"]),
            Paragraph("Кол-во", self._styles["table_header"]),
            Paragraph("Ед.", self._styles["table_header"]),
            Paragraph("Цена", self._styles["table_header"]),
            Paragraph("Сумма", self._styles["table_header"]),
        ]

    def _merge_works_by_name_price(self, works):
        """Merge rows with the same work name and price inside one table."""
        merged = {}
        order = []
        for work in works:
            name = work.get("name", "")
            price = int(work.get("price", 0))
            key = (name, price)
            if key not in merged:
                merged[key] = {
                    "name": name,
                    "price": price,
                    "quantity": float(work.get("quantity", 1)),
                    "unit": work.get("unit", ""),
                }
                order.append(key)
                continue
            merged[key]["quantity"] += float(work.get("quantity", 1))
        return [merged[key] for key in order]

    def _work_table_row(self, work):
        price = float(work.get("price", 0)) / 100.0
        quantity = float(work.get("quantity", 1))
        line_total = price * quantity
        unit = work.get("unit", "")
        return [
            Paragraph(self._escape(work.get("name", "")), self._styles["table_cell"]),
            Paragraph(self._escape(f"{quantity:g}"), self._styles["table_cell"]),
            Paragraph(self._escape(unit), self._styles["table_cell"]),
            Paragraph(self._escape(_format_money(price)), self._styles["table_cell"]),
            Paragraph(self._escape(_format_money(line_total)), self._styles["table_cell"]),
        ], line_total

    def _work_table_total_row(self, total):
        return [
            Paragraph("Итого:", self._styles["table_total"]),
            Paragraph("", self._styles["table_total"]),
            Paragraph("", self._styles["table_total"]),
            Paragraph("", self._styles["table_total"]),
            Paragraph(self._escape(_format_money(total)), self._styles["table_total"]),
        ]

    def _append_work_table(self, story, works, include_total_row=False):
        data = [self._work_table_header()]
        total = 0.0
        for work in self._merge_works_by_name_price(works):
            row, line_total = self._work_table_row(work)
            data.append(row)
            total += line_total

        if include_total_row:
            data.append(self._work_table_total_row(total))

        table = Table(data, colWidths=self._TABLE_COL_WIDTHS, repeatRows=1)
        table.setStyle(self._table_style(include_total_row))
        story.append(table)
        return total

    def _escape(self, value):
        text = str(value or "")
        return (
            text.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
        )

    def _paragraph(self, text, style_name="normal"):
        return Paragraph(self._escape(text), self._styles[style_name])

    def _append_header(self, story, options):
        if not options.get("include_report_header", True):
            return

        header_text = (options.get("report_header_text") or "").strip()
        if not header_text:
            header_text = "ОТЧЁТ О ПРОДЕЛАННЫХ РАБОТАХ"

        for line in header_text.splitlines() or [header_text]:
            story.append(self._paragraph(line, "title"))

        story.append(Spacer(1, 2 * mm))
        story.append(HRFlowable(width="100%", thickness=1, color=colors.black))
        story.append(Spacer(1, 4 * mm))

    def _meta_column_body(self, lines):
        if not lines:
            return self._paragraph("—")
        body = "<br/>".join(self._escape(line) for line in lines)
        return Paragraph(body, self._styles["normal"])

    def _meta_layout_style(self):
        return TableStyle([
            ("VALIGN", (0, 0), (-1, -1), "TOP"),
            ("LEFTPADDING", (0, 0), (-1, -1), 0),
            ("RIGHTPADDING", (0, 0), (0, -1), 8),
            ("TOPPADDING", (0, 0), (-1, -1), 0),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 0),
        ])

    def _append_meta(self, story, personal_info, client, object_data, options):
        story.append(Spacer(1, 6 * mm))
        
        if options.get("include_report_date", True):
            now = datetime.now().strftime("%d.%m.%Y %H:%M")
            story.append(self._paragraph(f"Дата формирования: {now}"))
            story.append(Spacer(1, 3 * mm))

        show_personal = options.get("include_personal_info", True)
        show_client = options.get("include_client_name", True)
        show_address = options.get("include_client_address", True)
        if not show_personal and not show_client and not show_address:
            return

        left_lines = []
        if show_personal:
            text = (personal_info or "").strip() or "—"
            left_lines = text.splitlines() or [text]

        right_lines = []
        if show_client:
            name = (client.get("name") or "").strip()
            if name:
                right_lines.append(name)
        if show_address:
            address = (object_data or {}).get("address") or client.get("address", "")
            if address:
                right_lines.append(f"Адрес: {address}")
            object_name = (object_data or {}).get("name", "")
            if object_name:
                right_lines.append(f"Объект: {object_name}")

        data = [
            [
                self._paragraph("Исполнитель:", "section"),
                self._paragraph("Заказчик:", "section"),
            ],
            [
                self._meta_column_body(left_lines),
                self._meta_column_body(right_lines),
            ],
        ]
        table = Table(data, colWidths=self._META_COL_WIDTHS)
        table.setStyle(self._meta_layout_style())
        story.append(table)
        story.append(Spacer(1, 4 * mm))

    def _append_works(self, story, works, options):
        story.append(Spacer(1, 2 * mm))

        if not works:
            story.append(self._paragraph("Нет записей о выполненных работах."))
            return

        total = 0.0
        if options.get("group_by_subobjects", True):
            groups = defaultdict(list)
            for work in works:
                key = (work.get("subobject_name") or "").strip()
                groups[key].append(work)

            for subobject_name in sorted(groups.keys(), key=str.casefold):
                group_works = groups[subobject_name]
                story.append(Spacer(1, 2 * mm))
                if subobject_name != '':
                    story.append(self._paragraph(f"{subobject_name}:", "section"))
                total += self._append_work_table(story, group_works)
        else:
            story.append(Spacer(1, 2 * mm))
            total = self._append_work_table(story, works, include_total_row=True)
            return

        story.append(Spacer(1, 2 * mm))
        total_table = Table(
            [self._work_table_total_row(total)],
            colWidths=self._TABLE_COL_WIDTHS,
        )
        total_table.setStyle(TableStyle([
            ("GRID", (0, 0), (-1, -1), 0.5, colors.grey),
            ("BACKGROUND", (0, 0), (-1, -1), colors.whitesmoke),
            ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
            ("TOPPADDING", (0, 0), (-1, -1), 3),
            ("BOTTOMPADDING", (0, 0), (-1, -1), 3),
            ("ALIGN", (0, 0), (3, 0), "RIGHT"),
            ("SPAN", (0, 0), (3, 0)),
        ]))
        story.append(total_table)

    def build_story(self, personal_info, client, object_data, works, options=None):
        """Return reportlab flowables for the report body."""
        options = options or {}
        story = []

        self._append_header(story, options)
        self._append_works(story, works, options)
        self._append_meta(story, personal_info, client, object_data, options)

        story.append(Spacer(1, 6 * mm))
        # story.append(HRFlowable(width="100%", thickness=1, color=colors.black))
        return story

    def build_pdf_bytes(self, personal_info, client, object_data, works, options=None):
        """Build PDF document and return bytes."""
        buffer = io.BytesIO()
        doc = SimpleDocTemplate(
            buffer,
            pagesize=A4,
            leftMargin=18 * mm,
            rightMargin=18 * mm,
            topMargin=18 * mm,
            bottomMargin=18 * mm,
            title="Work report",
        )
        doc.build(self.build_story(personal_info, client, object_data, works, options))
        return buffer.getvalue()

    def save_work_report(
        self,
        personal_info,
        client,
        object_data,
        works,
        options=None,
        file_path="",
    ):
        """Save PDF report to file and return path and bytes."""
        pdf_bytes = self.build_pdf_bytes(
            personal_info,
            client,
            object_data,
            works,
            options,
        )
        reports_dir = default_reports_dir()
        reports_dir.mkdir(parents=True, exist_ok=True)

        if file_path:
            target = Path(file_path).expanduser().resolve()
        else:
            safe_name = "".join(
                ch if ch.isalnum() else "_" for ch in client.get("name", "client")
            )
            stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            target = reports_dir / f"report_{safe_name}_{stamp}.pdf"

        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(pdf_bytes)
        return str(target), pdf_bytes


def save_work_report_pdf(
    personal_info,
    client,
    object_data,
    works,
    options=None,
    file_path="",
    font_path="",
):
    """Save PDF report using default builder."""
    builder = WorkReportPdfBuilder(font_path=font_path)
    return builder.save_work_report(
        personal_info,
        client,
        object_data,
        works,
        options,
        file_path=file_path,
    )

File: cpp/utils/pdfReportBuilder.cpp
#include "pdfReportBuilder.h"

#include "appPaths.h"
#include "databaseStorage.h"
#include "reportBuilder.h"

#include <QBuffer>
#include <QCoreApplication>
#include <QDateTime>
#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QMap>
#include <algorithm>
#include <QPageLayout>
#include <QPageSize>
#include <QPdfWriter>
#include <QTextDocument>
#include <QUrl>

namespace workorder
{

namespace
{

bool optionEnabled(const StringMap &options, const char *key, bool defaultValue = true)
{
    if(!options.contains(QLatin1String(key)))
        return defaultValue;
    return options.value(QLatin1String(key)).toBool();
}

QString safeClientName(const StringMap &client)
{
    const QString name = client.value("name").toString();
    QString safe;
    for(const QChar ch : name)
        safe.append(ch.isLetterOrNumber() ? ch : QChar('_'));
    return safe.isEmpty() ? QStringLiteral("client") : safe;
}

} // namespace

QString PdfReportBuilder::resolveReportFontPath(const QString &fontPath)
{
    if(!fontPath.isEmpty())
    {
        const QFileInfo info(fontPath);
        if(info.isFile())
            return info.absoluteFilePath();
    }

    const QString bundled = QStringLiteral(":/resources/fonts/DejaVuSans.ttf");
    if(QFileInfo::exists(bundled))
        return bundled;

    static const QStringList candidates = {
        "/System/Library/Fonts/Supplemental/Arial Unicode.ttf",
        "/Library/Fonts/Arial Unicode.ttf",
        "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
        "/usr/share/fonts/TTF/DejaVuSans.ttf",
        "C:/Windows/Fonts/arial.ttf",
    };

    for(const QString &candidate : candidates)
    {
        if(QFileInfo::exists(candidate))
            return candidate;
    }

    return QString();
}

QString PdfReportBuilder::registerReportFont(const QString &fontPath)
{
    const QString path = resolveReportFontPath(fontPath);
    if(path.isEmpty())
        return QStringLiteral("DejaVu Sans");

    return QFileInfo(path).completeBaseName();
}

PdfReportBuilder::PdfReportBuilder(const QString &fontPath)
    : m_fontPath(resolveReportFontPath(fontPath))
    , m_fontFamily(registerReportFont(m_fontPath))
{
}

QString PdfReportBuilder::escapeHtml(const QString &value) const
{
    QString text = value;
    text.replace('&', "&amp;");
    text.replace('<', "&lt;");
    text.replace('>', "&gt;");
    return text;
}

QString PdfReportBuilder::formatMoney(double value) const
{
    return QString::number(value, 'f', 2);
}

QString PdfReportBuilder::buildHtml(const QString &personalInfo, const StringMap &client,
                                    const StringMap &objectData, const StringMapList &works,
                                    const StringMap &options) const
{
    QString html = QStringLiteral("<html><head><meta charset=\"utf-8\"/>");
    if(!m_fontPath.isEmpty())
    {
        html += QStringLiteral("<style>@font-face { font-family: '%1'; src: url('%2'); }"
                               "body, table { font-family: '%1', sans-serif; font-size: 10pt; }</style>")
            .arg(escapeHtml(m_fontFamily), QUrl::fromLocalFile(m_fontPath).toString());
    }
    html += QStringLiteral("</head><body>");

    if(optionEnabled(options, "include_report_header"))
    {
        QString headerText = options.value("report_header_text").toString().trimmed();
        if(headerText.isEmpty())
            headerText = QStringLiteral("ОТЧЁТ О ПРОДЕЛАННЫХ РАБОТАХ");

        for(const QString &line : headerText.split('\n'))
            html += QStringLiteral("<h2 style=\"text-align:center;\">%1</h2>").arg(escapeHtml(line));

        html += QStringLiteral("<hr/>");
    }

    if(optionEnabled(options, "include_report_date"))
    {
        html += QStringLiteral("<p>Дата формирования: %1</p>")
            .arg(escapeHtml(QDateTime::currentDateTime().toString("dd.MM.yyyy HH:mm")));
    }

    if(optionEnabled(options, "include_personal_info")
        || optionEnabled(options, "include_client_name")
        || optionEnabled(options, "include_client_address"))
    {
        html += QStringLiteral("<table width=\"100%\"><tr><td valign=\"top\" width=\"50%\">");
        if(optionEnabled(options, "include_personal_info"))
        {
            html += QStringLiteral("<h3>Исполнитель:</h3><p>%1</p>")
                .arg(escapeHtml(personalInfo.trimmed().isEmpty() ? QStringLiteral("—") : personalInfo.trimmed()));
        }
        html += QStringLiteral("</td><td valign=\"top\" width=\"50%\">");
        if(optionEnabled(options, "include_client_name") || optionEnabled(options, "include_client_address"))
        {
            html += QStringLiteral("<h3>Заказчик:</h3>");
            if(optionEnabled(options, "include_client_name"))
                html += QStringLiteral("<p>%1</p>").arg(escapeHtml(client.value("name").toString()));
            if(optionEnabled(options, "include_client_address"))
            {
                QString address = objectData.value("address").toString();
                if(address.isEmpty())
                    address = client.value("address").toString();
                if(!address.isEmpty())
                    html += QStringLiteral("<p>Адрес: %1</p>").arg(escapeHtml(address));
                const QString objectName = objectData.value("name").toString();
                if(!objectName.isEmpty())
                    html += QStringLiteral("<p>Объект: %1</p>").arg(escapeHtml(objectName));
            }
        }
        html += QStringLiteral("</td></tr></table>");
    }

    html += QStringLiteral("<h3>Перечень работ</h3>");

    if(works.isEmpty())
    {
        html += QStringLiteral("<p>Нет записей о выполненных работах.</p>");
    }
    else
    {
        auto appendTable = [&](const StringMapList &rows, bool includeTotal)
        {
            double total = 0.0;
            html += QStringLiteral("<table border=\"1\" cellspacing=\"0\" cellpadding=\"4\" width=\"100%\">");
            html += QStringLiteral("<tr><th>Работа</th><th>Кол-во</th><th>Ед.</th><th>Цена</th><th>Сумма</th></tr>");

            QMap<QString, StringMap> merged;
            QStringList order;
            for(const QVariant &item : rows)
            {
                const StringMap work = item.toMap();
                const QString key = work.value("name").toString() + '|' + work.value("price").toString();
                if(!merged.contains(key))
                {
                    merged.insert(key, work);
                    order.append(key);
                    continue;
                }

                StringMap existing = merged.value(key);
                existing.insert("quantity", existing.value("quantity").toDouble() + work.value("quantity", 1).toDouble());
                merged.insert(key, existing);
            }

            for(const QString &key : order)
            {
                const StringMap work = merged.value(key);
                const double price = work.value("price").toDouble() / 100.0;
                const double quantity = work.value("quantity", 1).toDouble();
                const double lineTotal = price * quantity;
                total += lineTotal;

                html += QStringLiteral("<tr><td>%1</td><td align=\"right\">%2</td><td align=\"center\">%3</td>"
                                       "<td align=\"right\">%4</td><td align=\"right\">%5</td></tr>")
                    .arg(escapeHtml(work.value("name").toString()))
                    .arg(quantity, 0, 'g', 12)
                    .arg(escapeHtml(work.value("unit").toString()))
                    .arg(formatMoney(price))
                    .arg(formatMoney(lineTotal));
            }

            if(includeTotal)
            {
                html += QStringLiteral("<tr><td colspan=\"4\" align=\"right\"><b>Итого:</b></td>"
                                       "<td align=\"right\"><b>%1</b></td></tr>").arg(formatMoney(total));
            }

            html += QStringLiteral("</table>");
            return total;
        };

        double grandTotal = 0.0;
        if(optionEnabled(options, "group_by_subobjects"))
        {
            QMap<QString, StringMapList> groups;
            for(const QVariant &item : works)
            {
                const StringMap work = item.toMap();
                QString key = work.value("subobject_name").toString().trimmed();
                groups[key].append(work);
            }

            QStringList keys = groups.keys();
            std::sort(keys.begin(), keys.end(), [](const QString &a, const QString &b)
            {
                return a.compare(b, Qt::CaseInsensitive) < 0;
            });

            for(const QString &subobjectName : keys)
            {
                if(!subobjectName.isEmpty())
                    html += QStringLiteral("<h4>%1:</h4>").arg(escapeHtml(subobjectName));
                grandTotal += appendTable(groups.value(subobjectName), false);
            }

            html += QStringLiteral("<p align=\"right\"><b>Итого: %1</b></p>").arg(formatMoney(grandTotal));
        }
        else
        {
            appendTable(works, true);
        }
    }

    html += QStringLiteral("</body></html>");
    return html;
}

QByteArray PdfReportBuilder::buildPdfBytes(const QString &personalInfo, const StringMap &client,
                                           const StringMap &objectData, const StringMapList &works,
                                           const StringMap &options) const
{
    QTextDocument document;
    document.setHtml(buildHtml(personalInfo, client, objectData, works, options));

    QByteArray bytes;
    QBuffer buffer(&bytes);
    buffer.open(QIODevice::WriteOnly);

    QPdfWriter writer(&buffer);
    writer.setPageSize(QPageSize(QPageSize::A4));
    writer.setPageMargins(QMarginsF(18, 18, 18, 18), QPageLayout::Millimeter);
    document.print(&writer);

    return bytes;
}

QPair<QString, QByteArray> PdfReportBuilder::saveWorkReport(const QString &personalInfo, const StringMap &client,
                                                            const StringMap &objectData, const StringMapList &works,
                                                            const StringMap &options, const QString &filePath) const
{
    const QByteArray pdfBytes = buildPdfBytes(personalInfo, client, objectData, works, options);
    const QString reportsDir = DatabaseStorage::defaultReportsDir();
    QDir().mkpath(reportsDir);

    QString target;
    if(!filePath.isEmpty())
    {
        target = QFileInfo(filePath).absoluteFilePath();
    }
    else
    {
        const QString stamp = QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss");
        target = QDir(reportsDir).filePath(QStringLiteral("report_%1_%2.pdf").arg(safeClientName(client), stamp));
    }

    QDir().mkpath(QFileInfo(target).absolutePath());
    QFile file(target);
    if(file.open(QIODevice::WriteOnly))
        file.write(pdfBytes);

    return {target, pdfBytes};
}

QPair<QString, QByteArray> saveWorkReportPdf(const QString &personalInfo, const StringMap &client,
                                             const StringMap &objectData, const StringMapList &works,
                                             const StringMap &options, const QString &filePath,
                                             const QString &fontPath)
{
    return PdfReportBuilder(fontPath).saveWorkReport(personalInfo, client, objectData, works, options, filePath);
}

} // namespace workorder
